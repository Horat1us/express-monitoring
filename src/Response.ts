export enum State {
    ok = "ok",
    error = "error",
}

export interface ExecutionTime {
    begin: number; // unix timestamp, seconds
    total: number; // seconds
}

export interface Details {
    [k: string]: any;
}

export interface ErrorDetails<TDetails = Details> {
    type: string;
    code: number;
    message: string;
    details?: TDetails;
}

export interface BaseResponse {
    state: State;
    ms: ExecutionTime;
}

export interface SuccessResponse<TDetails = Details> extends BaseResponse {
    state: State.ok;
    details?: TDetails;
}

export interface FailureResponse<TDetails = Details> extends BaseResponse {
    state: State.error;
    error: ErrorDetails<TDetails>;
}

export type Response = SuccessResponse | FailureResponse;
