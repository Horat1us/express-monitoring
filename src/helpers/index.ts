export * from "./ExecutionTimer";
export * from "./getUnixTime";
export * from "./onFulfilled";
export * from "./onRejected";
export * from "./resolve";
