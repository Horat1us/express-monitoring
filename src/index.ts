import * as controls from "./controls";

export * from "./Response";
export * from "./FailureError";
export * from "./Controller";
export * from "./Control";
export { controls };
