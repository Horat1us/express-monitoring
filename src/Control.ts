import { Details } from "./Response";

export type Control = () => Promise<Details> | Details | undefined;
