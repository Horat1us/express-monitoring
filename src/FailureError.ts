import { Details, ErrorDetails } from "./Response";

export class FailureError<TDetails = Details> extends Error implements ErrorDetails<TDetails> {
    public readonly code: number;
    public readonly type: string;
    public readonly details?: TDetails;

    constructor(message: string, code: number, details?: TDetails, type?: string) {
        super(message);

        this.code = code;
        this.details = details;
        this.type = type || this.constructor.name;
    }

    public toRaw(): ErrorDetails<TDetails> {
        const error: ErrorDetails<TDetails> =  { code: this.code, type: this.type,  message: this.message };

        if (this.details) {
            error.details = this.details;
        }

        return error;
    }
}
