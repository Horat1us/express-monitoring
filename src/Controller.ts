import * as express from "express";
import { Control } from "./Control";
import * as helpers from "./helpers";
import * as controls from "./controls";

export interface ControllerConfig {
    controls: Array<{ id: string, control: Control }> | { [ k: string ]: Control } | Map<string, Control>;
}

export class Controller {
    public controls: Map<string, Control>;

    constructor(config: ControllerConfig) {
        this.controls = new Map<string, Control>();

        if (Array.isArray(config.controls)) {
            config.controls.forEach(
                ({ id, control }) => this.controls.set(id, control)
            );
        } else if (config.controls instanceof Map) {
            this.controls = config.controls;
        } else {
            Object.keys(config.controls).forEach(
                (id) => this.controls.set(id, (config.controls as { [ k: string ]: Control })[ id ])
            );
        }
    }

    public readonly control = async (req: express.Request, res: express.Response) => {
        if (!req.params.id) {
            return res.status(400).json({
                name: "Bad Request",
                message: "Missing required params: id.",
                code: 0,
                status: 400,
            });
        }

        const id: string = req.params.id;
        if (!this.controls.has(id)) {
            return res.status(404).json({
                name: "Not Found",
                message: `Control ${id} not found.`,
                code: -2,
            });
        }

        const control = this.controls.get(id) as Control;

        return this.execute(res, control);
    }

    public readonly full = async (req: express.Request, res: express.Response) => {
        const control = controls.Compose(this.controls);
        return this.execute(res, control);
    }

    private readonly execute = async (res: express.Response, control: Control) => {
        const response = await helpers.resolve(control);
        return res.status(200).json(response);
    }
}
