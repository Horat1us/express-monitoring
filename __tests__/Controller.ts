import express from "express";
import { Request } from 'jest-express/lib/request';
import { Response } from 'jest-express/lib/response';
import { Controller, FailureError, State } from "../src";

jest.mock('express', () => {
    return require('jest-express');
});

describe("Controller", () => {

    describe("controls configuration", () => {
        test("array of objects { id, control }", () => {
            const control = jest.fn();
            const controls = [
                {
                    id: "A",
                    control,
                },
            ];
            const controller = new Controller({ controls });

            expect(controller.controls.size).toEqual(controls.length);
            expect(controller.controls.get("A")).toEqual(control);
        });

        test("map", () => {
            const controls = new Map;
            const controller = new Controller({ controls });

            expect(controller.controls).toEqual(controls);
        });

        test("object { [id]: control }", () => {
            const control = jest.fn();
            const controls = {
                A: control,
            };
            const controller = new Controller({ controls });

            expect(controller.controls.size).toEqual(1);
            expect(controller.controls.get("A")).toEqual(control);
        });
    });

    describe("control handler", () => {

        let controller: Controller;
        let response: express.Response;
        beforeAll(() => {
            controller = new Controller({
                controls: new Map,
            });
            response = new Response as any;
        });

        test("returns 400 if missing id", async () => {
            const request = new Request("/monitoring");

            await controller.control(request as any, response as any);

            expect(response.status).toHaveBeenCalledWith(400);
            expect(response.json).toMatchSnapshot();
        });

        test("returns 404 if requested control not found", async () => {
            const request = new Request("/monitoring");
            request.setParams("id", "controlID");

            await controller.control(request as any, response);

            expect(response.status).toHaveBeenCalledWith(404);
            expect(response.json).toMatchSnapshot();
        });

        test("returns control successful response", async () => {
            const controlId = "A";
            const details = {
                key: Math.random(),
            };
            const request = new Request("/monitoring");
            request.setParams("id", controlId);
            controller.controls.set(controlId, () => (details));

            await controller.control(request as any, response);

            expect(response.status).toHaveBeenCalledWith(200);
            expect(response.json).toHaveBeenCalledTimes(1);
            expect((response.json as jest.Mock).mock.calls[ 0 ][ 0 ]).toMatchObject({
                state: State.ok,
                details,
            });
        });

        test("returns control failure response", async () => {
            const controlId = "A";
            const details = {
                key: Math.random(),
            };
            const failureError = new FailureError("Failure", 1, details, "TestFailure");
            const request = new Request("/monitoring");
            request.setParams("id", controlId);
            controller.controls.set(controlId, () => Promise.reject(failureError));

            await controller.control(request as any, response);

            expect(response.status).toHaveBeenCalledWith(200);
            expect(response.json).toHaveBeenCalledTimes(1);
            expect((response.json as jest.Mock).mock.calls[ 0 ][ 0 ]).toMatchObject({
                state: State.error,
                error: failureError.toRaw(),
            });
        });
    });

    describe("full handler", () => {

        let controller: Controller;
        let response: express.Response;
        beforeAll(() => {
            controller = new Controller({
                controls: new Map,
            });
            response = new Response as any;
        });

        test("returns compose control response", async () => {
            const resultA = "Result A";
            const resultB = new FailureError("Failure", 1, "Details");
            const controlA = jest.fn().mockResolvedValue(resultA);
            const controlB = jest.fn().mockResolvedValue(Promise.reject(resultB));

            controller.controls.set("A", controlA);
            controller.controls.set("B", controlB);

            await controller.full(new Request("/monitoring/full") as any, response);

            expect(response.status).toHaveBeenCalledWith(200);
            expect(response.json).toHaveBeenCalledTimes(1);
            expect((response.json as jest.Mock).mock.calls[0][0]).toMatchObject({
                state: State.error,
                error: {
                    details: {
                        A: {
                            state: State.ok,
                            details: resultA,
                        },
                        B: {
                            state: State.error,
                            error: {
                                details: resultB.details,
                            }
                        }
                    }
                }
            });
        });
    });
});
