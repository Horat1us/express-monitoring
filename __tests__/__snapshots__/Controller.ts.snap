// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller control handler returns 400 if missing id 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "code": 0,
        "message": "Missing required params: id.",
        "name": "Bad Request",
        "status": 400,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Controller control handler returns 404 if requested control not found 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "code": -2,
        "message": "Control controlID not found.",
        "name": "Not Found",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
